diff --git a/src/src/node_options.cc b/src/src/node_options.cc
index cd537ad68..47a63b6c0 100644
--- a/src/src/node_options.cc
+++ b/src/src/node_options.cc
@@ -729,6 +729,8 @@ PerIsolateOptionsParser::PerIsolateOptionsParser(
             V8Option{}, kAllowedInEnvironment);
   AddOption("--max-old-space-size", "", V8Option{}, kAllowedInEnvironment);
   AddOption("--max-semi-space-size", "", V8Option{}, kAllowedInEnvironment);
+  AddOption("--max-executable-size", "", V8Option{}, kAllowedInEnvironment);
+  AddOption("--max-heap-size", "", V8Option{}, kAllowedInEnvironment);
   AddOption("--perf-basic-prof", "", V8Option{}, kAllowedInEnvironment);
   AddOption("--perf-basic-prof-only-functions",
             "",
