--- a/src/deps/uv/uv.gyp
+++ b/src/deps/uv/uv.gyp
@@ -244,7 +244,7 @@
             'src/unix/sysinfo-memory.c',
           ],
           'link_settings': {
-            'libraries': [ '-ldl', '-lrt' ],
+            'libraries': [ '-ldl', '-lrt', '-latomic' ],
           },
         }],
         [ 'OS=="android"', {
--- a/src/deps/v8/gypfiles/v8.gyp
+++ b/src/deps/v8/gypfiles/v8.gyp
@@ -2031,7 +2031,8 @@
             'link_settings': {
               'libraries': [
                 '-ldl',
-                '-lrt'
+                '-lrt',
+                '-latomic'
               ],
             },
             'sources': [