diff --git a/src/tools/v8_gypfiles/toolchain.gypi b/src/tools/v8_gypfiles/toolchain.gypi
index 26c6866fa..9aa76e35b 100644
--- a/src/tools/v8_gypfiles/toolchain.gypi
+++ b/src/tools/v8_gypfiles/toolchain.gypi
@@ -763,7 +763,7 @@
               },
             },
             'variables': {
-              'v8_enable_slow_dchecks%': 1,
+              'v8_enable_slow_dchecks%': 0,
             },
             'conditions': [
               ['OS in "linux freebsd openbsd netbsd qnx aix os400 openharmony"', {
