diff --git a/src/node.gyp b/src/node.gyp
index e8e1d9f93..3e0782436 100644
--- a/src/node.gyp
+++ b/src/node.gyp
@@ -203,12 +203,12 @@
           'dependencies': [ '<(node_lib_target_name)' ],
           'conditions': [
             ['OS=="win" and node_shared=="true"', {
-              'dependencies': ['generate_node_def'],
-              'msvs_settings': {
-                'VCLinkerTool': {
-                  'ModuleDefinitionFile': '<(PRODUCT_DIR)/<(node_core_target_name).def',
-                },
-              },
+
+
+
+
+
+
             }],
           ],
         }],
@@ -1446,40 +1446,5 @@
         },
       ]
     }], # end aix section
-    ['OS=="win" and node_shared=="true"', {
-     'targets': [
-       {
-         'target_name': 'gen_node_def',
-         'type': 'executable',
-         'sources': [
-           'tools/gen_node_def.cc'
-         ],
-       },
-       {
-         'target_name': 'generate_node_def',
-         'dependencies': [
-           'gen_node_def',
-           '<(node_lib_target_name)',
-         ],
-         'type': 'none',
-         'actions': [
-           {
-             'action_name': 'generate_node_def_action',
-             'inputs': [
-               '<(PRODUCT_DIR)/<(node_lib_target_name).dll'
-             ],
-             'outputs': [
-               '<(PRODUCT_DIR)/<(node_core_target_name).def',
-             ],
-             'action': [
-               '<(PRODUCT_DIR)/gen_node_def.exe',
-               '<@(_inputs)',
-               '<@(_outputs)',
-             ],
-           },
-         ],
-       },
-     ],
-   }], # end win section
   ], # end conditions block
 }
