diff --git a/src/tools/v8_gypfiles/toolchain.gypi b/src/tools/v8_gypfiles/toolchain.gypi
index 26c6866fa..9aa76e35b 100644
--- a/src/tools/v8_gypfiles/toolchain.gypi
+++ b/src/tools/v8_gypfiles/toolchain.gypi
@@ -740,7 +740,7 @@
               },
             },
             'variables': {
-              'v8_enable_slow_dchecks%': 1,
+              'v8_enable_slow_dchecks%': 0,
             },
             'conditions': [
               ['OS=="linux" or OS=="freebsd" or OS=="openbsd" or OS=="netbsd" or \
