diff --git a/src/node.gyp b/src/node.gyp
index 4e614df1a..c5b3fa96a 100644
--- a/src/node.gyp
+++ b/src/node.gyp
@@ -1332,7 +1372,7 @@
            {
              'action_name': 'generate_node_def_action',
              'inputs': [
-               '<(PRODUCT_DIR)/<(node_lib_target_name).dll'
+               '<(PRODUCT_DIR)/<(node_core_target_name).dll'
              ],
              'outputs': [
                '<(PRODUCT_DIR)/<(node_core_target_name).def',
