diff --git a/src/deps/undici/undici.js b/src/deps/undici/undici.js
index cccf3f104..6967976a6 100644
--- a/src/deps/undici/undici.js
+++ b/src/deps/undici/undici.js
@@ -2301,6 +2301,7 @@ var require_pool_base = __commonJS({
 var require_timers = __commonJS({
   "lib/util/timers.js"(exports2, module2) {
     "use strict";
+    const {setTimeout, clearTimeout, setInterval, clearInterval } = require('timers');
     var fastNow = 0;
     var RESOLUTION_MS = 1e3;
     var TICK_MS = (RESOLUTION_MS >> 1) - 1;
@@ -2346,7 +2347,7 @@ var require_timers = __commonJS({
       } else {
         clearTimeout(fastNowTimeout);
         fastNowTimeout = setTimeout(onTick, TICK_MS);
-        if (fastNowTimeout.unref) {
+        if (fastNowTimeout && fastNowTimeout.unref) {
           fastNowTimeout.unref();
         }
       }
