diff --git a/src/deps/v8/src/diagnostics/unwinding-info-win64.cc b/src/deps/v8/src/diagnostics/unwinding-info-win64.cc
index a71b86613..5eb4c3f10 100644
--- a/src/deps/v8/src/diagnostics/unwinding-info-win64.cc
+++ b/src/deps/v8/src/diagnostics/unwinding-info-win64.cc
@@ -21,6 +21,36 @@
 // This has to come after windows.h.
 #include <versionhelpers.h>  // For IsWindows8OrGreater().
 
+// Forward declaration to keep this independent of Win8
+NTSYSAPI
+DWORD
+NTAPI
+RtlAddGrowableFunctionTable(
+    _Out_ PVOID* DynamicTable,
+    _In_reads_(MaximumEntryCount) PRUNTIME_FUNCTION FunctionTable,
+    _In_ DWORD EntryCount,
+    _In_ DWORD MaximumEntryCount,
+    _In_ ULONG_PTR RangeBase,
+    _In_ ULONG_PTR RangeEnd
+    );
+
+
+NTSYSAPI
+void
+NTAPI
+RtlGrowFunctionTable(
+    _Inout_ PVOID DynamicTable,
+    _In_ DWORD NewEntryCount
+    );
+
+
+NTSYSAPI
+void
+NTAPI
+RtlDeleteGrowableFunctionTable(
+    _In_ PVOID DynamicTable
+    );
+
 namespace v8 {
 namespace internal {
 namespace win64_unwindinfo {
