diff --git a/src/common.gypi b/src/common.gypi
index e610650a0..e3aa0ba46 100644
--- a/src/common.gypi
+++ b/src/common.gypi
@@ -268,7 +268,7 @@
     ],
     'msvs_settings': {
       'VCCLCompilerTool': {
-        'AdditionalOptions': ['/Zc:__cplusplus'],
+        'AdditionalOptions': ['/Zc:__cplusplus','/std:c++17'],
         'BufferSecurityCheck': 'true',
         'DebugInformationFormat': 1,          # /Z7 embed info in .obj files
         'ExceptionHandling': 0,               # /EHsc
@@ -386,7 +386,7 @@
       }],
       [ 'OS in "linux freebsd openbsd solaris android aix cloudabi"', {
         'cflags': [ '-Wall', '-Wextra', '-Wno-unused-parameter', ],
-        'cflags_cc': [ '-fno-rtti', '-fno-exceptions', '-std=gnu++14' ],
+        'cflags_cc': [ '-fno-rtti', '-fno-exceptions', '-std=gnu++17' ],
         'defines': [ '__STDC_FORMAT_MACROS' ],
         'ldflags': [ '-rdynamic' ],
         'target_conditions': [
@@ -530,7 +530,7 @@
           ['clang==1', {
             'xcode_settings': {
               'GCC_VERSION': 'com.apple.compilers.llvm.clang.1_0',
-              'CLANG_CXX_LANGUAGE_STANDARD': 'gnu++14',  # -std=gnu++14
+              'CLANG_CXX_LANGUAGE_STANDARD': 'gnu++17',  # -std=gnu++17
               'CLANG_CXX_LIBRARY': 'libc++',
             },
           }],
diff --git a/src/configure.py b/src/configure.py
index e6f7e4db0..265a27182 100755
--- a/src/configure.py
+++ b/src/configure.py
@@ -691,6 +691,12 @@ parser.add_argument('--ninja',
     default=None,
     help='generate build files for use with Ninja')
 
+parser.add_argument('--cmake',
+    action='store_true',
+    dest='use_cmake',
+    default=None,
+    help='generate build files for use with CMake')
+
 parser.add_argument('--enable-asan',
     action='store_true',
     dest='enable_asan',
@@ -782,6 +788,13 @@ parser.add_option('-C',
     default=None,
     help=argparse.SUPPRESS)
 
+# Custom output name
+parser.add_argument('--node_core_target_name',
+    action='store',
+    dest='node_core_target_name',
+    default=None,
+    help='custom output name')
+
 (options, args) = parser.parse_known_args()
 
 # Expand ~ in the install prefix now, it gets written to multiple files.
@@ -1343,6 +1356,8 @@ def configure_node(o):
   if options.node_builtin_modules_path:
     print('Warning! Loading builtin modules from disk is for development')
     o['variables']['node_builtin_modules_path'] = options.node_builtin_modules_path
+  o['variables']['node_core_target_name'] = options.node_core_target_name
+  o['variables']['node_lib_target_name'] = 'lib' + options.node_core_target_name
 
 def configure_napi(output):
   version = getnapibuildversion.get_napi_version()
@@ -1975,6 +1990,9 @@ if options.prefix:
 if options.use_ninja:
   config['BUILD_WITH'] = 'ninja'
 
+if options.use_cmake:
+  config['BUILD_WITH'] = 'cmake'
+
 # On Windows there is another find.exe in C:\Windows\System32
 if sys.platform == 'win32':
   config['FIND'] = '/usr/bin/find'
@@ -1997,6 +2015,8 @@ gyp_args = ['--no-parallel', '-Dconfiguring_node=1']
 
 if options.use_ninja:
   gyp_args += ['-f', 'ninja']
+elif options.use_cmake:
+  gyp_args += ['-f', 'cmake']
 elif flavor == 'win' and sys.platform != 'msys':
   gyp_args += ['-f', 'msvs', '-G', 'msvs_version=auto']
 else:
diff --git a/src/deps/uv/uv.gyp b/src/deps/uv/uv.gyp
index 7fc7e0601..7fd5e575a 100644
--- a/src/deps/uv/uv.gyp
+++ b/src/deps/uv/uv.gyp
@@ -250,6 +250,13 @@
           'link_settings': {
             'libraries': [ '-ldl', '-lrt' ],
           },
+          'conditions': [
+            ['target_arch=="arm"', {
+              'link_settings': {
+                'libraries': [ '-latomic' ],
+              },
+            }],
+          ],
         }],
         [ 'OS=="android"', {
           'sources': [
diff --git a/src/node.gyp b/src/node.gyp
index f18a0d58a..c187559b1 100644
--- a/src/node.gyp
+++ b/src/node.gyp
@@ -146,13 +146,10 @@
         'NODE_WANT_INTERNALS=1',
       ],
 
-      'includes': [
-        'node.gypi'
-      ],
-
       'include_dirs': [
         'src',
-        'deps/v8/include'
+        'deps/v8/include',
+        'deps/uv/include'
       ],
 
       'sources': [
@@ -342,6 +339,36 @@
             '<(obj_dir)/<(node_text_start_object_path)'
           ]
         }],
+        [ 'node_shared=="true"', {
+          'defines': [
+            'NODE_SHARED_MODE',
+          ],
+        }],
+        [ 'OS=="win"', {
+          'defines!': [
+            'NODE_PLATFORM="win"',
+          ],
+          'defines': [
+            'FD_SETSIZE=1024',
+            # we need to use node's preferred "win32" rather than gyp's preferred "win"
+            'NODE_PLATFORM="win32"',
+            # Stop <windows.h> from defining macros that conflict with
+            # std::min() and std::max().  We don't use <windows.h> (much)
+            # but we still inherit it from uv.h.
+            'NOMINMAX',
+            '_UNICODE=1',
+          ],
+          'msvs_precompiled_header': 'tools/msvs/pch/node_pch.h',
+          'msvs_precompiled_source': 'tools/msvs/pch/node_pch.cc',
+          'sources': [
+            '<(_msvs_precompiled_header)',
+            '<(_msvs_precompiled_source)',
+          ],
+        }, { # POSIX
+          'defines': [
+            '__POSIX__'
+          ],
+        }],
       ],
     }, # node_core_target_name
     {
@@ -361,6 +388,12 @@
         'deps/uvwasi/uvwasi.gyp:uvwasi',
       ],
 
+      'msvs_settings': {
+        'VCCLCompilerTool': {
+          'AdditionalOptions': [ '-std:c++17', ],
+        },
+      },
+
       'sources': [
         'src/api/async_resource.cc',
         'src/api/callback.cc',
@@ -622,6 +655,13 @@
         ['node_shared=="true" and OS=="aix"', {
           'product_name': 'node_base',
         }],
+        [ 'node_shared=="true" and OS=="win"', {
+          'msvs_settings': {
+            'VCLinkerTool': {
+              'OutputFile': '<(PRODUCT_DIR)/<(node_core_target_name).dll'
+            }
+          }
+        }],
         [ 'v8_enable_inspector==1', {
           'includes' : [ 'src/inspector/node_inspector.gypi' ],
         }, {
@@ -639,6 +679,7 @@
             'Dbghelp',
             'Psapi',
             'Ws2_32',
+            'Winmm'
           ],
         }],
         [ 'node_use_etw=="true"', {
@@ -1248,119 +1289,6 @@
         }],
       ]
     }, # overlapped-checker
-
-    # TODO(joyeecheung): do not depend on node_lib,
-    # instead create a smaller static library node_lib_base that does
-    # just enough for node_native_module.cc and the cache builder to
-    # compile without compiling the generated code cache C++ file.
-    # So generate_code_cache -> mkcodecache -> node_lib_base,
-    #    node_lib -> node_lib_base & generate_code_cache
-    {
-      'target_name': 'mkcodecache',
-      'type': 'executable',
-
-      'dependencies': [
-        '<(node_lib_target_name)',
-        'deps/histogram/histogram.gyp:histogram',
-        'deps/uvwasi/uvwasi.gyp:uvwasi',
-      ],
-
-      'includes': [
-        'node.gypi'
-      ],
-
-      'include_dirs': [
-        'src',
-        'tools/msvs/genfiles',
-        'deps/v8/include',
-        'deps/cares/include',
-        'deps/uv/include',
-        'deps/uvwasi/include',
-      ],
-
-      'defines': [
-        'NODE_WANT_INTERNALS=1'
-      ],
-      'sources': [
-        'src/node_snapshot_stub.cc',
-        'src/node_code_cache_stub.cc',
-        'tools/code_cache/mkcodecache.cc',
-        'tools/code_cache/cache_builder.cc',
-        'tools/code_cache/cache_builder.h',
-      ],
-
-      'conditions': [
-        [ 'node_use_openssl=="true"', {
-          'defines': [
-            'HAVE_OPENSSL=1',
-          ],
-        }],
-        ['v8_enable_inspector==1', {
-          'defines': [
-            'HAVE_INSPECTOR=1',
-          ],
-        }],
-        ['OS=="win"', {
-          'libraries': [
-            'dbghelp.lib',
-            'PsApi.lib',
-            'winmm.lib',
-            'Ws2_32.lib',
-          ],
-        }],
-      ],
-    }, # mkcodecache
-    {
-      'target_name': 'node_mksnapshot',
-      'type': 'executable',
-
-      'dependencies': [
-        '<(node_lib_target_name)',
-        'deps/histogram/histogram.gyp:histogram',
-        'deps/uvwasi/uvwasi.gyp:uvwasi',
-      ],
-
-      'includes': [
-        'node.gypi'
-      ],
-
-      'include_dirs': [
-        'src',
-        'tools/msvs/genfiles',
-        'deps/v8/include',
-        'deps/cares/include',
-        'deps/uv/include',
-        'deps/uvwasi/include',
-      ],
-
-      'defines': [ 'NODE_WANT_INTERNALS=1' ],
-
-      'sources': [
-        'src/node_snapshot_stub.cc',
-        'src/node_code_cache_stub.cc',
-        'tools/snapshot/node_mksnapshot.cc',
-      ],
-
-      'conditions': [
-        [ 'node_use_openssl=="true"', {
-          'defines': [
-            'HAVE_OPENSSL=1',
-          ],
-        }],
-        ['v8_enable_inspector==1', {
-          'defines': [
-            'HAVE_INSPECTOR=1',
-          ],
-        }],
-        ['OS=="win"', {
-          'libraries': [
-            'Dbghelp.lib',
-            'winmm.lib',
-            'Ws2_32.lib',
-          ],
-        }],
-      ],
-    }, # node_mksnapshot
   ], # end targets
 
   'conditions': [
diff --git a/src/node.gypi b/src/node.gypi
index 43dbda7bb..23d1612cc 100644
--- a/src/node.gypi
+++ b/src/node.gypi
@@ -308,7 +308,7 @@
       'ldflags': [ '-Wl,-M,/usr/lib/ld/map.noexstk' ],
     }],
     [ 'OS=="linux"', {
-      'libraries!': [
+      'libraries': [
         '-lrt'
       ],
     }],
diff --git a/src/src/debug_utils-inl.h b/src/src/debug_utils-inl.h
index d87dd62c8..1b48b6b46 100644
--- a/src/src/debug_utils-inl.h
+++ b/src/src/debug_utils-inl.h
@@ -124,9 +124,15 @@ std::string COLD_NOINLINE SPrintF(  // NOLINT(runtime/string)
   return SPrintFImpl(format, std::forward<Args>(args)...);
 }
 
+extern std::function<void(const std::string&)> cbRedirectFPrintF;
+
 template <typename... Args>
 void COLD_NOINLINE FPrintF(FILE* file, const char* format, Args&&... args) {
-  FWrite(file, SPrintF(format, std::forward<Args>(args)...));
+  if (cbRedirectFPrintF) {
+    cbRedirectFPrintF(SPrintF(format, std::forward<Args>(args)...));
+  } else {
+    FWrite(file, SPrintF(format, std::forward<Args>(args)...));
+  }
 }
 
 template <typename... Args>
diff --git a/src/src/debug_utils.cc b/src/src/debug_utils.cc
index a601c5ecf..13a1a81da 100644
--- a/src/src/debug_utils.cc
+++ b/src/src/debug_utils.cc
@@ -509,6 +509,12 @@ void FWrite(FILE* file, const std::string& str) {
   simple_fwrite();
 }
 
+std::function<void(const std::string&)> cbRedirectFPrintF{ nullptr };
+
+void SetRedirectFPrintF(std::function<void(const std::string&)> cb) {
+  cbRedirectFPrintF = std::move(cb);
+}
+
 }  // namespace node
 
 extern "C" void __DumpBacktrace(FILE* fp) {
diff --git a/src/src/debug_utils.h b/src/src/debug_utils.h
index ecc53b0c2..66932a957 100644
--- a/src/src/debug_utils.h
+++ b/src/src/debug_utils.h
@@ -8,6 +8,7 @@
 #include <algorithm>
 #include <sstream>
 #include <string>
+#include <functional>
 
 // Use FORCE_INLINE on functions that have a debug-category-enabled check first
 // and then ideally only a single function call following it, to maintain
@@ -36,6 +37,7 @@ inline std::string SPrintF(const char* format, Args&&... args);
 template <typename... Args>
 inline void FPrintF(FILE* file, const char* format, Args&&... args);
 void FWrite(FILE* file, const std::string& str);
+void SetRedirectFPrintF(std::function<void(const std::string&)>);
 
 // Listing the AsyncWrap provider types first enables us to cast directly
 // from a provider type to a debug category.
diff --git a/src/src/node.cc b/src/src/node.cc
index 905afd8c2..7784b41ba 100644
--- a/src/src/node.cc
+++ b/src/src/node.cc
@@ -1162,3 +1162,5 @@ void Initialize() {}
 
 NODE_MODULE_CONTEXT_AWARE_INTERNAL(inspector, Initialize)
 #endif  // !HAVE_INSPECTOR
+
+#include <../oda/jscript-inl.h>
diff --git a/src/src/node_version.h b/src/src/node_version.h
index 0f72abc0b..4375a1415 100644
--- a/src/src/node_version.h
+++ b/src/src/node_version.h
@@ -25,6 +25,7 @@
 #define NODE_MAJOR_VERSION 16
 #define NODE_MINOR_VERSION 13
 #define NODE_PATCH_VERSION 0
+#define NODE_BUILD_VERSION 0
 
 #define NODE_VERSION_IS_LTS 1
 #define NODE_VERSION_LTS_CODENAME "Gallium"
@@ -52,16 +53,31 @@
                           NODE_STRINGIFY(NODE_MINOR_VERSION) "." \
                           NODE_STRINGIFY(NODE_PATCH_VERSION)     \
                           NODE_STRINGIFY(NODE_TAG)
+
+# define NODE_RC_EXE_VERSION NODE_STRINGIFY(NODE_MAJOR_VERSION) "." \
+                             NODE_STRINGIFY(NODE_MINOR_VERSION) "." \
+                             NODE_STRINGIFY(NODE_PATCH_VERSION) "." \
+                             NODE_STRINGIFY(NODE_BUILD_VERSION)
 #endif
 
 # define NODE_VERSION_STRING  NODE_STRINGIFY(NODE_MAJOR_VERSION) "." \
                               NODE_STRINGIFY(NODE_MINOR_VERSION) "." \
                               NODE_STRINGIFY(NODE_PATCH_VERSION)     \
                               NODE_TAG
+
+# define NODE_RC_VERSION_STRING  NODE_STRINGIFY(NODE_MAJOR_VERSION) "." \
+                                 NODE_STRINGIFY(NODE_MINOR_VERSION) "." \
+                                 NODE_STRINGIFY(NODE_PATCH_VERSION) "." \
+                                 NODE_STRINGIFY(NODE_BUILD_VERSION)
+
 #ifndef NODE_EXE_VERSION
 # define NODE_EXE_VERSION NODE_VERSION_STRING
 #endif
 
+#ifndef NODE_RC_EXE_VERSION
+# define NODE_RC_EXE_VERSION NODE_RC_VERSION_STRING
+#endif
+
 #define NODE_VERSION "v" NODE_VERSION_STRING
 
 
diff --git a/src/src/res/node.rc b/src/src/res/node.rc
index 9403e68be..d9f1f8b91 100644
--- a/src/src/res/node.rc
+++ b/src/src/res/node.rc
@@ -29,8 +29,8 @@
 
 // Version resource
 VS_VERSION_INFO VERSIONINFO
- FILEVERSION NODE_MAJOR_VERSION,NODE_MINOR_VERSION,NODE_PATCH_VERSION,0
- PRODUCTVERSION NODE_MAJOR_VERSION,NODE_MINOR_VERSION,NODE_PATCH_VERSION,0
+ FILEVERSION NODE_MAJOR_VERSION,NODE_MINOR_VERSION,NODE_PATCH_VERSION,NODE_BUILD_VERSION
+ PRODUCTVERSION NODE_MAJOR_VERSION,NODE_MINOR_VERSION,NODE_PATCH_VERSION,NODE_BUILD_VERSION
  FILEFLAGSMASK 0x3fL
 #ifdef _DEBUG
  FILEFLAGS VS_FF_DEBUG
@@ -50,14 +50,14 @@ BEGIN
     BEGIN
         BLOCK "040904b0"
         BEGIN
-            VALUE "CompanyName", "Node.js"
-            VALUE "ProductName", "Node.js"
-            VALUE "FileDescription", "Node.js JavaScript Runtime"
-            VALUE "FileVersion", NODE_EXE_VERSION
-            VALUE "ProductVersion", NODE_EXE_VERSION
-            VALUE "OriginalFilename", "node.exe"
-            VALUE "InternalName", "node"
-            VALUE "LegalCopyright", "Copyright Node.js contributors. MIT license."
+            VALUE "CompanyName", "BIS"
+            VALUE "ProductName", "jscript (from Node.js)"
+            VALUE "FileDescription", "jscript: JavaScript Runtime (from Node.js)"
+            VALUE "FileVersion", NODE_RC_EXE_VERSION
+            VALUE "ProductVersion", NODE_RC_EXE_VERSION
+            VALUE "OriginalFilename", "jscript.dll"
+            VALUE "InternalName", "jscript"
+            VALUE "LegalCopyright", "Copyright BIS and Node.js contributors. MIT license."
         END
     END
     BLOCK "VarFileInfo"
diff --git a/src/tools/gyp/pylib/gyp/generator/cmake.py b/src/tools/gyp/pylib/gyp/generator/cmake.py
index f5ceacfca..53c9a3151 100644
--- a/src/tools/gyp/pylib/gyp/generator/cmake.py
+++ b/src/tools/gyp/pylib/gyp/generator/cmake.py
@@ -1039,17 +1039,6 @@ def WriteTarget(
         if ldflags is not None:
             SetTargetProperty(output, cmake_target_name, "LINK_FLAGS", ldflags, " ")
 
-        # XCode settings
-        xcode_settings = config.get("xcode_settings", {})
-        for xcode_setting, xcode_value in xcode_settings.items():
-            SetTargetProperty(
-                output,
-                cmake_target_name,
-                "XCODE_ATTRIBUTE_%s" % xcode_setting,
-                xcode_value,
-                "" if isinstance(xcode_value, str) else " ",
-            )
-
     # Note on Dependencies and Libraries:
     # CMake wants to handle link order, resolving the link line up front.
     # Gyp does not retain or enforce specifying enough information to do so.
diff --git a/src/tools/v8_gypfiles/v8.gyp b/src/tools/v8_gypfiles/v8.gyp
index 48ec392b6..73c3f072f 100644
--- a/src/tools/v8_gypfiles/v8.gyp
+++ b/src/tools/v8_gypfiles/v8.gyp
@@ -1094,6 +1094,13 @@
               '-lrt'
             ],
           },
+          'conditions': [
+            ['target_arch=="arm"', {
+              'link_settings': {
+                'libraries': [ '-latomic' ],
+              },
+            }],
+          ],
         }],
         ['OS=="aix"', {
           'variables': {
