diff --git a/src/common.gypi b/src/common.gypi
index e610650a0..e3aa0ba46 100644
--- a/src/common.gypi
+++ b/src/common.gypi
@@ -268,6 +268,7 @@
         'SuppressStartupBanner': 'true',
         'WarnAsError': 'false',
         'WarningLevel': 3,                    # /W3
+        'AdditionalOptions': [ '/std:c++17', ],
       },
       'VCLinkerTool': {
         'target_conditions': [
@@ -374,7 +375,7 @@
       }],
       [ 'OS in "linux freebsd openbsd solaris android aix cloudabi"', {
         'cflags': [ '-Wall', '-Wextra', '-Wno-unused-parameter', ],
-        'cflags_cc': [ '-fno-rtti', '-fno-exceptions', '-std=gnu++1y' ],
+        'cflags_cc': [ '-fno-rtti', '-fno-exceptions', '-std=gnu++17' ],
         'defines': [ '__STDC_FORMAT_MACROS' ],
         'ldflags': [ '-rdynamic' ],
         'target_conditions': [
@@ -519,7 +520,7 @@
           ['clang==1', {
             'xcode_settings': {
               'GCC_VERSION': 'com.apple.compilers.llvm.clang.1_0',
-              'CLANG_CXX_LANGUAGE_STANDARD': 'gnu++1y',  # -std=gnu++1y
+              'CLANG_CXX_LANGUAGE_STANDARD': 'gnu++17',  # -std=gnu++17
               'CLANG_CXX_LIBRARY': 'libc++',
             },
           }],
