diff --git a/src/lib/internal/modules/esm/resolve.js b/src/lib/internal/modules/esm/resolve.js
index dadad4ff57..5500908d6e 100644
--- a/src/lib/internal/modules/esm/resolve.js
+++ b/src/lib/internal/modules/esm/resolve.js
@@ -927,6 +927,37 @@ function packageResolve(specifier, base, conditions) {
     // Cross-platform root check.
   } while (packageJSONPath.length !== lastPath.length);
 
+  if (CJSModule && CJSModule.globalPaths && Array.isArray(CJSModule.globalPaths)) {
+    for (const path of CJSModule.globalPaths) {
+      if (!path) {
+          continue;
+      }
+      const baseUrl =
+        path.startsWith('file:') ? path
+                                 : 'file://' + path;
+      const packageJSONUrl =
+        new URL(baseUrl + '/' + packageName + '/package.json');
+
+      const packageJSONPath = fileURLToPath(packageJSONUrl);
+
+      const stat = tryStatSync(StringPrototypeSlice(packageJSONPath, 0,
+                                                    packageJSONPath.length - 13));
+      if (!stat.isDirectory()) {
+        continue;
+      }
+
+      // Package match.
+      const packageConfig = getPackageConfig(packageJSONPath, specifier, base);
+      if (packageConfig.exports !== undefined && packageConfig.exports !== null)
+        return packageExportsResolve(
+          packageJSONUrl, packageSubpath, packageConfig, base, conditions
+        ).resolved;
+      if (packageSubpath === '.')
+        return legacyMainResolve(packageJSONUrl, packageConfig, base);
+      return new URL(packageSubpath, packageJSONUrl);
+    }
+  }
+
   // eslint can't handle the above code.
   // eslint-disable-next-line no-unreachable
   throw new ERR_MODULE_NOT_FOUND(packageName, fileURLToPath(base));
