diff --git a/src/deps/v8/src/wasm/function-body-decoder-impl.h b/src/deps/v8/src/wasm/function-body-decoder-impl.h
index 48b804a3a..03829319c 100644
--- a/src/deps/v8/src/wasm/function-body-decoder-impl.h
+++ b/src/deps/v8/src/wasm/function-body-decoder-impl.h
@@ -2905,15 +2905,17 @@ class WasmFullDecoder : public WasmDecoder<validate> {
                      br_arity, startrel(control_.back().pc), available);
         return false;
       }
-      Value* stack_values = &*(stack_.end() - br_arity);
-      // Type-check the topmost br_arity values on the stack.
-      for (int i = 0; i < br_arity; ++i) {
-        Value& val = stack_values[i];
-        if (!val.type.IsSubTypeOf(result_types[i])) {
-          this->errorf(this->pc_,
-                       "type error in merge[%u] (expected %s, got %s)", i,
-                       result_types[i].type_name(), val.type.type_name());
-          return false;
+      if (br_arity > 0) {
+        Value* stack_values = &*(stack_.end() - br_arity);
+        // Type-check the topmost br_arity values on the stack.
+        for (int i = 0; i < br_arity; ++i) {
+          Value& val = stack_values[i];
+          if (!val.type.IsSubTypeOf(result_types[i])) {
+            this->errorf(this->pc_,
+              "type error in merge[%u] (expected %s, got %s)", i,
+              result_types[i].type_name(), val.type.type_name());
+            return false;
+          }
         }
       }
     } else {  // !control_.back().reachable()
